src/aspire/source/__init__.py

```diff
@@ -370,6 +373,151 @@ def whiten(self, noise_filter):
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = (images * np.expand_dims(signal_mask, 2))
+            noise = (images * np.expand_dims(noise_mask, 2))
+            signal_mean += np.sum(signal)
+            noise_mean += np.sum(noise)
+        signal_denominator = self.n * np.sum(signal_mask)
+        noise_denominator = self.n * np.sum(noise_mask)
+        signal_mean /= signal_denominator
+        noise_mean /= noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to reverse contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to reverse contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, bg_radius=1.0, single_img=True, do_ramp=True,
+                             batch_size=512):
+        """
+        Normalize the images by the noise background
+
+        This is done by shifting the image density by the mean value of background
+        and scaling the image density by the standard deviation of background.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending the `Add` and `Multiple` filters to the generation pipeline.
+
+        :param bg_radius: Radius cutoff to be considered as background (in image size)
+        :param single_img: When it is `True`, normalize background based on the mean
+            and std values from corresponding individual images.
+        :param do_ramp: When it is `True`, fit a ramping background to the data
+            and subtract. Namely perform normalization based on values from each image.
+            Otherwise, a constant background level from all images is used.
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+
+        logger.info(f'Normalize background on source object with radius size of {bg_radius}')
+        L = self.L
+        grid = grid_2d(L)
+        mask = (grid['r'] > bg_radius)
+
+        if do_ramp:
+            # Create matrices and reshape the background mask
+            # for fitting a ramping background
+            ramp_mask = np.vstack((grid['x'][mask].flatten(),
+                                   grid['y'][mask].flatten(),
+                                   np.ones(grid['y'][mask].flatten().size))).T
+            ramp_all = np.vstack((grid['x'].flatten(), grid['y'].flatten(),
+                                  np.ones(L*L))).T
+            mask_reshape = mask.reshape((L*L))
+
+        if single_img:
```

Comment:
There is a lot of duplication here between the two branches (`single_img == True` and `single_img == False`). It should be possible to reduce this by properly using broadcasting and calculating a mean or two somewhere if `single_img == False`.

That being said, it's not clear to me why we need the `single_img == False` case. Is there any documented use of this in the code? As far as I can tell, this is more or less handles by `whiten` since that normalizes the noise power spectrum to be unity. It might be simpler right now to remove the option and always do per-image normalization.

Reply:
@junchaoxia .. there are more problems. `mask_reshape` is undefined if `do_ramp` is False. A unit test would have revealed this issue. Let's add one today after we go through what's happening here.

Reply:
Thanks. The mask_reshape was removed when I reimplemented the do_ramp preprocess as pipeline in the new version.